python/lsst/ts/salobj/base.py

```diff
@@ -230,37 +245,169 @@ def tai_from_utc(utc, format="unix"):
     """
-    if isinstance(utc, astropy.time.Time):
-        ap_time = utc
+    if isinstance(utc, float) and format == "unix":
+        utc_unix = utc
+    elif isinstance(utc, astropy.time.Time):
+        utc_unix = utc.unix
     else:
-        ap_time = astropy.time.Time(utc, scale="utc", format=format)
-    return ap_time.tai.mjd * SECONDS_PER_DAY - MJD_MINUS_UNIX_SECONDS
+        utc_unix = astropy.time.Time(utc, scale="utc", format=format).unix
+    return tai_from_utc_unix(utc_unix)
+
+
+def tai_from_utc_unix(utc_unix):
+    """Return TAI in unix seconds, given UTC in unix seconds.
+
+    Parameters
+    ----------
+    utc_unix : `float`
+        UTC time in unix seconds.
+
+    Returns
+    -------
+    tai_unix : `float`
+        TAI time in unix seconds.
+
+    Raises
+    ------
+    ValueError
+        If the date is earlier than 1972 (which is before integer leap seconds)
+        or within one day of the expiration date of the leap second table
+        (which is automatically updated).
+
+    Notes
+    -----
+    See the notes for `tai_from_utc` for information about
+    possibly unexpected behavior on the day before a leap second.
+    """
+    # Use a local pointer, to prevent race conditions while the
+    # global table is being replaced by `_update_leap_second_table`.
+    leap_second_table = _LEAP_SECOND_TABLE
+    if utc_unix > leap_second_table[-1][0] - SECONDS_PER_DAY:
+        raise ValueError(
+            f"{utc_unix} > expiry date of leap second table - 1 day "
+            f"= {leap_second_table[-1][0] - SECONDS_PER_DAY}"
+        )
+    i = bisect.bisect(leap_second_table, (utc_unix, math.inf))
+    if i == 0:
+        raise ValueError(
+            f"{utc_unix} < start of integer leap seconds "
+            f"= {leap_second_table[0][0]}"
+        )
+    utc0, tai_minus_utc0 = leap_second_table[i - 1]
+    utc1, tai_minus_utc1 = leap_second_table[i]
+    if utc_unix + SECONDS_PER_DAY > utc1 and tai_minus_utc1 is not None:
+        # Assume unix seconds is smeared uniformly on the day before a
+        # leap second, so that there are exactly 86400 seconds in the day.
+        # Otherwise unix seconds is ambiguous at the leap second.
+        # This matches AstroPy and Standards of Fundamental Astronomy (SOFA).
+        utc_days = utc_unix / SECONDS_PER_DAY
+        frac_day = utc_days - math.floor(utc_days)
+        tai_minus_utc = tai_minus_utc0 + (tai_minus_utc1 - tai_minus_utc0) * frac_day
+    else:
+        tai_minus_utc = tai_minus_utc0
+    return utc_unix + tai_minus_utc
+
+
+_log = logging.getLogger("lsst.ts.salobj.base")
+
+
+def _update_leap_second_table():
+    """Update the leap second table.
+
+    Notes
+    -----
+    This should be called when this module is loaded.
+    When called, it obtains the current table from AstroPy,
+    then schedules a background (daemon) thread to call itself
+    to update the table ``_LEAP_SECOND_TABLE_UPDATE_MARGIN_DAYS``
+    before the table expires.
```

Comment:
So as I understand things, each salobj user is going to ask astropy to update the leap second table, have it written to the standard shared astropy cache as needed, and then read that table in as an internal leap second table?

Reply:
I would express it this way: every time a user imports salobj in Python, salobj it will ask AstroPy for a leap second table. What AstroPy will do depends on how it is configured:
* If AstroPy uses its own table, and its own table is out of date, it will download a new table. Otherwise it will return a copy of the cached table.
* If AstroPy is configured to use the table maintained by NTP or PTP then it will return that table. If that table is out of date we're in trouble, but I think we can choose whether to allow AstroPy to download a new table at that point.